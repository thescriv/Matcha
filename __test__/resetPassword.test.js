const crypto = require("crypto")
const superagent = require("superagent")
const dotenv = require("dotenv")

const db = require("../src/lib/db")

const { script } = require("../scripts/script")
const { deleteRows } = require("./utils/deleteRows")
const { transporter } = require("../src/lib/transporter")
const { getRandomPort } = require("./utils/getRandomPort")
const { startApi, stopApi } = require("../api")

let baseUrl

describe(`resetPassword --`, () => {
  beforeAll(async () => {
    const port = getRandomPort()

    dotenv.config({ path: "../.env" })

    baseUrl = `http://localhost:${port}/api`

    await script("matchaTest")

    await startApi(port)
  })

  beforeEach(() => {
    jest.spyOn(Date, "now").mockReturnValue(1000000000000)
  })

  afterEach(async () => {
    jest.restoreAllMocks()

    await deleteRows()
  })

  afterAll(async () => {
    await stopApi()
  })

  describe("POST /", () => {
    test("do send a resetPassword", async () => {
      jest.spyOn(transporter, "sendMail").mockReturnValue(true)
      jest
        .spyOn(crypto, "randomBytes")
        .mockReturnValue("0000000000000000000000000000000000000001")

      await db.query(
        `INSERT INTO user (nickname, firstname, lastname, email, password) 
            VALUES ("thescriv", "theo", "test", "test@test.test", "f94b937b40c946b916c63a1b638ee956")`
      )

      const { body, status } = await superagent
        .post(`${baseUrl}/resetPassword/`)
        .send({
          email: "test@test.test",
        })

      expect({ body, status }).toMatchSnapshot()

      const [[user], [resetPasswordToken]] = await Promise.all([
        db.query("SELECT * FROM user"),
        db.query("SELECT * FROM password_reset_token"),
      ])
      expect(user).toMatchSnapshot()
      expect(resetPasswordToken).toMatchSnapshot()
    })

    test("do not send a resetPassword (user does not exist)", async () => {
      let error

      try {
        await superagent.post(`${baseUrl}/resetPassword/`).send({
          email: "test@test.test",
        })
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()

      const [resetPasswordToken] = await db.query(
        "SELECT * FROM password_reset_token"
      )
      expect(resetPasswordToken).toMatchSnapshot()
    })

    test("do not send a resetPassword (sendEmail failed)", async () => {
      jest.spyOn(transporter, "sendMail").mockReturnValue(false)

      await db.query(
        `INSERT INTO user (nickname, firstname, lastname, email, password) 
            VALUES ("thescriv", "theo", "test", "test@test.test", "f94b937b40c946b916c63a1b638ee956")`
      )

      let error

      try {
        await superagent.post(`${baseUrl}/resetPassword/`).send({
          email: "test@test.test",
        })
      } catch (err) {
        error = err
      }

      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()

      const [resetPasswordToken] = await db.query(
        "SELECT * FROM password_reset_token"
      )
      expect(resetPasswordToken).toMatchSnapshot()
    })
  })

  describe("POST /:tokenId", () => {
    test("do reset password", async () => {
      jest.spyOn(transporter, "sendMail").mockReturnValue(true)
      jest
        .spyOn(crypto, "randomBytes")
        .mockReturnValue("0000000000000000000000000000000000000001")

      const {
        body: bodyRegisterUser,
        status: statusRegisterUser,
      } = await superagent.post(`${baseUrl}/register`).send({
        nickname: "thescriv",
        firstname: "theo",
        lastname: "test",
        email: "test@test.test",
        password: "testTest1",
      })

      expect({ bodyRegisterUser, statusRegisterUser })

      const {
        body: bodyResetPassword,
        status: statusResetPassword,
      } = await superagent.post(`${baseUrl}/resetPassword/`).send({
        email: "test@test.test",
      })

      expect({ bodyResetPassword, statusResetPassword })

      const { body, status } = await superagent
        .post(
          `${baseUrl}/resetPassword/0000000000000000000000000000000000000001`
        )
        .send({
          password: "password",
        })

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()

      const [resetPasswordToken] = await db.query(
        "SELECT * FROM password_reset_token"
      )
      expect(resetPasswordToken).toMatchSnapshot()
    })

    test("do not reset password (token does not exist)", async () => {
      await db.query(
        `INSERT INTO user (nickname, firstname, lastname, email, password) 
            VALUES ("thescriv", "theo", "test", "test@test.test", "f94b937b40c946b916c63a1b638ee956")`
      )

      let error

      try {
        await superagent
          .post(
            `${baseUrl}/resetPassword/0000000000000000000000000000000000000001`
          )
          .send({
            password: "password",
          })
      } catch (err) {
        error = err
      }
      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()

      const [resetPasswordToken] = await db.query(
        "SELECT * FROM password_reset_token"
      )
      expect(resetPasswordToken).toMatchSnapshot()
    })

    test("do not reset password (user does not exist)", async () => {
      await db.query(
        `INSERT INTO password_reset_token (user_id, token) VALUES ("1", "0000000000000000000000000000000000000001")`
      )
      let error

      try {
        await superagent
          .post(
            `${baseUrl}/resetPassword/0000000000000000000000000000000000000001`
          )
          .send({
            password: "password",
          })
      } catch (err) {
        error = err
      }
      const { body, status } = error.response

      expect({ body, status }).toMatchSnapshot()

      const [user] = await db.query("SELECT * FROM user")
      expect(user).toMatchSnapshot()

      const [resetPasswordToken] = await db.query(
        "SELECT * FROM password_reset_token"
      )
      expect(resetPasswordToken).toMatchSnapshot()
    })
  })
})
