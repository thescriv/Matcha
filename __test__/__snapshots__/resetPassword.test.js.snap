// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resetPassword -- POST / do not send a resetPassword (sendEmail failed) 1`] = `
Object {
  "body": Object {
    "error": "api.resetPassword sendMail failed",
  },
  "status": 400,
}
`;

exports[`resetPassword -- POST / do not send a resetPassword (sendEmail failed) 2`] = `
Array [
  TextRow {
    "bio": null,
    "completed": 0,
    "email": "test@test.test",
    "firstname": "theo",
    "gender_id": null,
    "id": 1,
    "lastname": "test",
    "nickname": "thescriv",
    "orientation_id": 1,
    "password": "f94b937b40c946b916c63a1b638ee956",
    "score": 100,
    "verified": 0,
  },
]
`;

exports[`resetPassword -- POST / do not send a resetPassword (sendEmail failed) 3`] = `Array []`;

exports[`resetPassword -- POST / do not send a resetPassword (user does not exist) 1`] = `
Object {
  "body": Object {
    "error": "api.resetPassword user does_not_exist",
  },
  "status": 400,
}
`;

exports[`resetPassword -- POST / do not send a resetPassword (user does not exist) 2`] = `Array []`;

exports[`resetPassword -- POST / do not send a resetPassword (user does not exist) 3`] = `Array []`;

exports[`resetPassword -- POST / do send a resetPassword 1`] = `
Object {
  "body": Object {},
  "status": 204,
}
`;

exports[`resetPassword -- POST / do send a resetPassword 2`] = `
Array [
  TextRow {
    "bio": null,
    "completed": 0,
    "email": "test@test.test",
    "firstname": "theo",
    "gender_id": null,
    "id": 1,
    "lastname": "test",
    "nickname": "thescriv",
    "orientation_id": 1,
    "password": "f94b937b40c946b916c63a1b638ee956",
    "score": 100,
    "verified": 0,
  },
]
`;

exports[`resetPassword -- POST / do send a resetPassword 3`] = `
Array [
  TextRow {
    "creation_date": 1000000000000,
    "id": 1,
    "token": "0000000000000000000000000000000000000001",
    "user_id": 1,
  },
]
`;

exports[`resetPassword -- POST /:tokenId do not reset password (token does not exist) 1`] = `
Object {
  "body": Object {
    "error": "api.resetPassword token does_not_exist",
  },
  "status": 400,
}
`;

exports[`resetPassword -- POST /:tokenId do not reset password (token does not exist) 2`] = `
Array [
  TextRow {
    "bio": null,
    "completed": 0,
    "email": "test@test.test",
    "firstname": "theo",
    "gender_id": null,
    "id": 1,
    "lastname": "test",
    "nickname": "thescriv",
    "orientation_id": 1,
    "password": "f94b937b40c946b916c63a1b638ee956",
    "score": 100,
    "verified": 0,
  },
]
`;

exports[`resetPassword -- POST /:tokenId do not reset password (token does not exist) 3`] = `Array []`;

exports[`resetPassword -- POST /:tokenId do not reset password (user does not exist) 1`] = `
Object {
  "body": Object {
    "error": "api.resetPassword token does_not_exist",
  },
  "status": 400,
}
`;

exports[`resetPassword -- POST /:tokenId do not reset password (user does not exist) 2`] = `Array []`;

exports[`resetPassword -- POST /:tokenId do not reset password (user does not exist) 3`] = `Array []`;

exports[`resetPassword -- POST /:tokenId do reset password 1`] = `
Object {
  "body": Object {},
  "status": 204,
}
`;

exports[`resetPassword -- POST /:tokenId do reset password 2`] = `
Array [
  TextRow {
    "bio": null,
    "completed": 0,
    "email": "test@test.test",
    "firstname": "theo",
    "gender_id": null,
    "id": 1,
    "lastname": "test",
    "nickname": "thescriv",
    "orientation_id": 1,
    "password": "5f4dcc3b5aa765d61d8327deb882cf99",
    "score": 100,
    "verified": 0,
  },
]
`;

exports[`resetPassword -- POST /:tokenId do reset password 3`] = `
Array [
  TextRow {
    "creation_date": 1000000000000,
    "id": 1,
    "token": "0000000000000000000000000000000000000001",
    "user_id": 1,
  },
]
`;
